client/transaction-pool/src/revalidation.rs

```diff
@@ -0,0 +1,346 @@
+				.or_insert_with(|| {
+					let mut bt = HashSet::new();
+					bt.insert(ext_hash.clone());
+					bt
+				});
+			self.members.insert(ext_hash.clone(), block_number);
+		}
+	}
+}
+
+impl<Api: ChainApi> Future for RevalidationWorker<Api>
+where
+	Api: 'static,
+{
+	type Output = ();
+
+	fn poll(mut self: Pin<&mut Self>, cx: &mut std::task::Context) -> Poll<Self::Output> {
+		let this = &mut *self;
+		let mut process_some_immediately = false;
+
+		while let Poll::Ready(msg) = this.from_queue.poll_next_unpin(cx) {
+			match msg {
+				Some(worker_payload) => {
+					this.best_block = worker_payload.at;
+					this.push(worker_payload);
+					process_some_immediately = true;
+					continue;
+				},
+				// R.I.P. worker!
+				None => return Poll::Ready(()),
+			}
+		}
+
+		loop {
+			match (this.tick_timeout.poll_next_unpin(cx), process_some_immediately) {
+				(Poll::Ready(Some(_)), _) | (_, true) => {
+					process_some_immediately = false;
+
+					let next_batch = this.prepare_batch();
+					let batch_revalidate_fut =
+						batch_revalidate(this.pool.clone(), this.api.clone(), this.best_block, next_batch);
+
+					futures::pin_mut!(batch_revalidate_fut);
+					match batch_revalidate_fut.poll_unpin(cx)
+					{
+						Poll::Ready(_) => continue,
+						_ => break,
+					}
+				},
+				_ => break,
+			}
+		}
+
+
+		Poll::Pending
+	}
+}
+
+/// Background state for revalidation queue.
+struct BackgroundConfig<Api: ChainApi> {
+	to_worker: mpsc::UnboundedSender<WorkerPayload<Api>>,
+	_spawner: futures::executor::ThreadPool,
+}
+
+/// Revalidation queue.
+///
+/// Can be configured background (`new_background`)
+/// or immediate (just `new`).
+pub struct RevalidationQueue<Api: ChainApi> {
+	pool: Arc<Pool<Api>>,
+	api: Arc<Api>,
+	background: Option<BackgroundConfig<Api>>,
+}
+
+impl<Api: ChainApi> RevalidationQueue<Api>
+where
+	Api: 'static,
+{
+	/// New revalidation queue without background worker.
+	pub fn new(api: Arc<Api>, pool: Arc<Pool<Api>>) -> Self {
+		Self {
+			api,
+			pool,
+			background: None,
+		}
+	}
+
+	/// New revalidation queue with background worker.
+	pub fn new_background(api: Arc<Api>, pool: Arc<Pool<Api>>) -> Self {
+		let spawner = futures::executor::ThreadPool::builder()
+			.name_prefix("txpool-worker")
+			.pool_size(1)
+			.create()
+			.expect("Creating worker thread task failed");
+
+		let (to_worker, from_queue) = mpsc::unbounded();
+
+		let worker = RevalidationWorker::new(api.clone(), pool.clone(), from_queue);
+		spawner.spawn_ok(worker);
+
+		Self {
+			api,
+			pool,
+			background: Some(
+				BackgroundConfig { _spawner: spawner, to_worker }
+			),
+		}
+	}
+
+	/// Offload some revalidation to the queue.
+	///
+	/// If queue configured with background worker, this will return immediately.
+	/// If queue configured without background worker, this will resolve after
+	/// revalidation is actually done.
+	pub async fn offload(&self, at: NumberFor<Api>, transactions: Vec<ExHash<Api>>) {
```

Comment:
I'd call it `revalidate_later`, `offload` seems pretty generic and it's hard to understand what we off-load without a lot of context.